### DISCUSSION TITLE
Feature requests (v2)

### DISCUSSION THREAD

**Discussion by Ni55aN** [retejs/rete#631] [Ideas]
By collecting feature requests and use case information directly from our users, we can gain insights into their pain points, identify common patterns and focus on developing features that will have the greatest impact.

To ensure effective communication and gather comprehensive information, we kindly request users to provide the following details in their feature requests (feature per comment):

1. **Feature Description**: Please provide a clear and concise description of the desired feature (2-3 sentences). This should include what the feature aims to accomplish and how it would enhance the user experience.
2. **Use Case Context**: Describe the specific context in which the feature would be used. Help us understand the problem you're trying to solve or the task you're trying to achieve. Providing real-world examples or scenarios where the feature would be beneficial will greatly assist us in assessing its potential value.
3. **Existing Solutions**: If you are aware of any existing solutions or workarounds that partially address your feature request, please share them with us. This will help us identify gaps in the current functionality and potentially leverage existing solutions as a starting point.

In return for submitting a feature request, you can expect one of the following from us, based on the nature of the feature:

- **General-Purpose Plugin**: If the requested feature aligns with the project's overall goals and do not require excessive configuring for many use cases, we will consider developing it as a plugin. This way, it can be easily integrated into the existing application without affecting its core functionality 
- **Example Implementation**: In cases where multiple solutions are possible, and the feature heavily depends on specific use cases or applications, we will provide an example. This will demonstrate how the feature can be implemented and customized based on your specific requirements.

⚠ If you would rather provide your feedback in a more informal manner or if you are unsure whether a similar functionality or workaround already exists, you can join our [Discord server](https://discord.gg/cxSFkPZdsV) and share your feature request there.

💡 Please use the "thumbs-up" 👍 reaction to show support for existing feature requests rather than duplicating them in separate comments.

### v1
You may be curious about the progress made on feature requests from the [previous version](https://github.com/retejs/rete/issues/80). Most of them had been actively reviewed based on your feedback

### Sponsorship and contribution

If you are interested in supporting the development of a specific feature, you can consider [sponsoring](https://retejs.org/sponsor) its implementation or providing drafts or prototypes.

**Comment by Ni55aN** [comment]
**Pan the viewport when dragging a node**

1. Enable panning when dragging a node near viewport edge. When the dragged node reaches the boundaries of the viewport, the area should start translating.
2. It would be helpful if users could drag a node beyond the boundaries of the viewport without having to release the node and separately pan the viewport
3. Unreal Engine Blueprint

![ue](https://github.com/retejs/rete/assets/8259641/b210f9ae-a8a9-4838-87ff-f4ea34a177db)

**Reply by Ni55aN** [reply]
Implemented in the [Panning boundary](https://retejs.org/examples/panning-boundary) example

**Comment by ShaitanLyss** [comment]
**Macro nodes**

1. Enable the creation of macro nodes, a macro node is a node that represents a collapsed node graph.
2. It allows users to create their own custom nodes for reusability and sharing
3. This could be implemented by creating a node editor for each macro node but I assume it would be slightly more optimal if they would all be part of the same node editor (maybe not) or much better if they could be compiled in some way (codegen ?). This feature request is akin to unreal blueprint macros.

**Reply by thomasmansencal** [reply]
Yes please, I know that there are scopes which are not too far from this but being able to group nodes in a reusable block is a very important feature for adoption!

**Comment by stallboy** [comment]
### Keep the node position and render it sync

1. store the node position with id, when called createEditor again, I can pass the position info in parameters and render it sync so the UI is not flickering.
2. the `createEditor` function transforms parameter to node and connection. that's simplest programming model when use rete, but it surfers from the flickering problem when I update the parameter the node position change and the node created async .
3. one can update the node and connection incremental, but the code is really complex, the mind model turns from pure function model to state-machine model.  the other problem is the how to auto layout it incremental..

![flash](https://github.com/retejs/rete/assets/559049/0e8c87e0-5fae-4eb1-84a7-485e458cf973)

**Comment by mrwhy-orig** [comment]
I'd like to be able to create the connections in a 90° manner. Like in wiring diagrams. If they cross each other ther should be a bridge or a connection point.

**Comment by egalvis39** [comment]
I was wondering if there is a possibility to add a functionality to copy/cut and paste nodes within the editor. I think this would be a very useful addition, especially when working with large diagrams or when we need to replicate nodes quickly and accurately. This could improve productivity and avoid errors when duplicating manually.

**Comment by aminasharaf** [comment]
I need to call a function when a node is deselected. Is there an event triggered upon node deselection

**Reply by Ni55aN** [reply]
it can be achieved by extending the selector https://retejs.org/docs/guides/selectable#extend-selector

**Comment by heitorgiacomini** [comment]
Can i hide input and change connection position based on the other node like in mysql Workbench?

![connection](https://github.com/user-attachments/assets/6e2665e0-a8b0-428d-a240-d915237f3bfa)

**Comment by tobyski-tdsultra** [comment]
I would like to be able to save the full graph at a readable zoom level to an image, I've played with html2canvas but only get either unreadable zoom level version or just a crop of the graph.

**Comment by pokecheater** [comment]
# An Interface inside the template files to obtain the Retejs plugins (like editor, area_plugin, connection_plugin).

Hi Retejs team,

I stumbled over the problem, that inside my CustomConnection.vue file the only properties I have are the conneciton proeprties itself, like source, target, sourceSocket, targetSocket, name etc.

But to give my connection the same color as the socket I would need to know of which type the socket itself is.
For that in turn I need to access the editor property. But that is not so easy, since there are no properties provided inside template itself.

By now I solve this with the following workaround:

![image](https://github.com/user-attachments/assets/755d9605-2ec0-4dab-ac54-0cb2b1dffcff)


Please note, that the object of ReteNodeEditor is my custom class which holds all the plugins and some useful helper functions:

![image](https://github.com/user-attachments/assets/5ff096a6-7e31-4bbf-9ccb-8f4ff3d8c6e9)

Then the problem is easy solveable:
![image](https://github.com/user-attachments/assets/2e713fab-51d5-4ddd-9b27-f0969848113c)


Maybe you already have such an interface but I was not able to find an example of it.

Best greetings and thanks in advance.

**Comment by proehlen** [comment]
Please provide a simple and elegant way to disable the "double click to zoom" functionality.  The user clicking the up or down buttons too quickly in a numeric input field resulting in a zoom is weird and very annoying.

Issue #444 provides a code work around but, as we can see from the issue, it is fragile accross versions and hardly elegant.

**Comment by Sathiskumar-S** [comment]
How to create a select as an input?
I want to add a node that contains select boxes as well as when the user selects a particular option it automatically passes the node value to connected. How to achieve that  using angular. As well as, how to export the nodes and connection as json file in recent rete.js?

**Comment by LAWArthur** [comment]
### A Filter for valid parents in the Scopes plugin

I find it hard to restrict the nodes that can be treated as parent. Presently, every node can become a parent, which is making the plugin quite unusable in practice.

By reading the source code, I'm pretty sure that currently this cannot be done through a pipe, a user-defined preset or something, but only through directly modifying the plugin. However, I regard it easy to add a filter as an injection, which can immediately solve the trouble and add much more flexibility.

**Comment by samerickson** [comment]
## undo/redo stack cap

I want to be able to cap the number of entries that get added to the history plugin to something sane like 50-100.

**Comment by TomCarpendale** [comment]
## Add an identifier to the context menu

For custom positioning of the context menu (e.g. to keep it clamped within the container), it would be useful to be able to query it by an id or attribute. Currently, there is a workaround by querying `'[data-testid="context-menu"]'`, but it should be a simple fix to add an id or attribute to the context menu element.

**Comment by SanjaRogovic** [comment]
Expose undo/redo availability state

The `HistoryPlugin` exposes `undo()` and `redo()`, but applications with toolbar buttons also need to know whether Undo and Redo are currently available.

For example, an Angular toolbar might have:

```html
<button [disabled]="!canUndo">Undo</button>
<button [disabled]="!canRedo">Redo</button>
```

Currently, the only way I found to determine this is by accessing the plugin's internal history state (`produced` and `reserved`), which relies on a private implementation detail.

It would be useful if the plugin exposed this through a public API, for example:

```ts
history.canUndo(): boolean;
history.canRedo(): boolean;
```

or

```ts
history.getState(): {
  canUndo: boolean;
  canRedo: boolean;
}
```

This would allow applications to enable or disable Undo/Redo actions without relying on internal properties.

**Comment by liruiyu330** [comment]
Dear Rete.js developer,

I am a developer and I'm using your framework to build a process editing tool. I've run into some issues during development and am writing to ask for your advice.

I'm using the Hybrid Engine model in Rete.js. When working with control flow, I've encountered a problem: the flowchart cannot clearly distinguish between branching and parallel execution. Both situations simply appear as one node with two outgoing connections to two other nodes.

I would like to ask whether there is already a well-established solution for this in the original design, or if you have any suggestions.

I look forward to your reply.

Thank you.

**Comment by samerickson** [comment]
Hello, I would like to request a feature request where the comment plugin is supported by the read only plugin. Thank you very much!
